VagrantでRedmineを構築する

kazuki-sumi opened the pull request:
```
vagrant up
ansible-playbook -i hosts redmine.yml
```
ansible playbook実行後、仮想環境にsshでアクセスすし、MySQLのユーザ設定をする(/var/log/mysqld.log)。初期パスワードがあるので、それを使いMySQLのrootユーザのパスワードを変更する。
```
mysql -u root -p 旧パスワード --connect-expired-password -e "ALTER USER 'root'@'localhost' IDENTIFIED BY 新パスワード;"
```
その後、redmineのdatabase.ymlのproduction用のDBのパスワードに変更後のものを記述。

/var/www/redmineに移動して下記を実行してください。
```
bundle install
bundle exec rake generate_secret_token
bundle exec rake db:create RAILS_ENV=production
bundle exec rake db:migrate RAILS_ENV=production
```
http://192.168.33.10/  にアクセスし、redmineの画面が表示されます。

kazuki-sumi committed:
centos環境構築

kazuki-sumi committed:
unicornのsystemd設定ファイル作成

kazuki-sumi committed:
nginxの設定

kazuki-sumi committed:
unicornの設定

kazuki-sumi committed:
unicorn起動コマンド修正

kazuki-sumi committed:
ansibleの設定

kazuki-sumi committed:
base-packageインストールタスク作成

kazuki-sumi committed:
vagrantのsudo権限でするように修正

kazuki-sumi committed:
nginxタスク作成

kazuki-sumi committed:
mysqlタスク作成

kazuki-sumi committed:
rubyインストールのタスク作成

kazuki-sumi committed:
rubyインストール先を/usr/local/に変更。パスの設定ファイルも/etc/profileに移動

kazuki-sumi committed:
/home/ユーザ/下での構築をまたする時用に保存

kazuki-sumi committed:
bundle install時に必要になったので追加

kazuki-sumi committed:
railsタスク作成

kazuki-sumi committed:
rubyビルド環境に必要なツールを追加

kazuki-sumi committed:
redmine構築タスク作成

kazuki-sumi committed:
unicornタスク作成

kazuki-sumi committed:
redmineタスク修正

RVIRUS0817 commented on redmine/group_vars/redmine:
```diff
@@ -0,0 +1,5 @@
+HOME: "/home/vagrant"
+# mysql
+MYSQL_VERSION: "http://dev.mysql.com/get/mysql57-community-release-el7-7.noarch.rpm"
+MYSQL_PASSWORD: "Password@1"
```
これはセキュリティ的に隠さないとダメですね。

RVIRUS0817 commented on redmine/redmine.conf:
```diff
@@ -0,0 +1,32 @@
+# unixドメイン:同じマシン上で動いているプロセスが通信を行うためのソケット。
+# リクエストをappサーバーに渡す時、サーバーグループ(今回はunicorn-redmine)を定義し、proxy_passで指定する。
+upstream unicorn-redmine {
+    server unix:/opt/redmine/tmp/unicorn_redmine.sock;
```
unicorn.sockでいいかと

RVIRUS0817 commented on redmine/redmine.conf:
```diff
@@ -0,0 +1,32 @@
+
+server {
+    listen 80;
+    server_name 192.168.33.10;
```
redmine-menta.meとかでアクセスできるようにすればいいですね

RVIRUS0817 commented on redmine/redmine.conf:
```diff
@@ -0,0 +1,32 @@
+    listen 80;
+    server_name 192.168.33.10;
+
+    root /opt/redmine/public;
```
/optよりも/var/www配下が一般的ですね

RVIRUS0817 commented on redmine/roles/mysql/tasks/main.yml:
```diff
@@ -0,0 +1,28 @@
+- name: update expired root user password
+  shell: mysql -u root -p"{{ MYSQL_DEFAULT_PASSWORD.stdout }}" --connect-expired-password -e "ALTER USER 'root'@'localhost' IDENTIFIED BY '{{ MYSQL_PASSWORD }}';"
+  ignore_errors: True
+  #changed_when: result.rc != true
```
この処理は初期設定なので、毎回そのコマンドが打たれてしまいますが、
冪等性的に問題ないですか？

RVIRUS0817 commented on redmine/roles/nginx/files/redmine.conf:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    error_page 500 502 503 504 /500.html;
+}
```
上記のredmine/redmine.confと重複してませんか？

RVIRUS0817 commented on redmine/roles/nginx/tasks/main.yml:
```diff
@@ -0,0 +1,25 @@
+- name: install nginx and openssl
+  yum: name={{ item }} state=present
+  loop:
+    - nginx
```
1つだけなら直で書けばコードが短くなります。

RVIRUS0817 commented on redmine/roles/rails/tasks/main.yml:
```diff
@@ -0,0 +1,7 @@
+  ignore_errors: yes
+
+- name: install rails 5.2.2
+  shell: bash -lc "gem install rails --version="5.2.2""
```
group_varsとかでバージョン管理するとわかりやすいかと

RVIRUS0817 commented on redmine/roles/redmine/files/Gemfile:
```diff
@@ -0,0 +1,102 @@
+      when /sqlite3/
+        gem "sqlite3", "~>1.3.12", :platforms => [:mri, :mingw, :x64_mingw]
+      when /sqlserver/
+        gem "tiny_tds", "~> 1.0.5", :platforms => [:mri, :mingw, :x64_mingw]
+        gem "activerecord-sqlserver-adapter", :platforms => [:mri, :mingw, :x64_mingw]
+      else
+        warn("Unknown database adapter `#{adapter}` found in config/database.yml, use Gemfile.local to load your own database gems")
+      end
+    end
+  else
+    warn("No adapter found in config/database.yml, please configure it first")
+  end
+else
+  warn("Please configure your config/database.yml first")
+end
+
+group :development do
+  gem "rdoc"
+  gem "yard"
+end
+
+group :test do
+  gem "rails-dom-testing"
+  gem "mocha"
+  gem "simplecov", "~> 0.14.1", :require => false
+  # For running system tests
+  gem 'puma', '~> 3.7'
+  gem "capybara", '~> 2.13'
+  gem "selenium-webdriver"
+end
+
+local_gemfile = File.join(File.dirname(__FILE__), "Gemfile.local")
+if File.exists?(local_gemfile)
+  eval_gemfile local_gemfile
+end
+
+# Load plugins' Gemfiles
+Dir.glob File.expand_path("../plugins/*/{Gemfile,PluginGemfile}", __FILE__) do |file|
+  eval_gemfile file
+end
```
Gemfileはansibleで管理というより他のリポジトリで管理したほうが良いと思いますよ。

kazuki-sumi replied:
今回はこれでOK

RVIRUS0817 commented on redmine/roles/redmine/tasks/main.yml:
```diff
@@ -0,0 +1,55 @@
+  command:
+    cmd: mv redmine-4.0.0 redmine
+    chdir: /opt/
+    creates: /opt/redmine
```
これだと毎回redmineがダウンロードされてしまうので
ファイルがあったらダウンロードしないという条件文が可能です。

RVIRUS0817 commented on redmine/roles/redmine/tasks/main.yml:
```diff
@@ -0,0 +1,55 @@
+- name: copy database.yml
+  copy:
+    dest: /opt/redmine/config
+    src: ./roles/redmine/files/database.yml
```
この `.` っていりますか？
相対パスでいける気がします

RVIRUS0817 commented on redmine/roles/unicorn/files/redmine-unicorn.service:
```diff
@@ -0,0 +1,20 @@
+ExecReload=/bin/kill -USR2 $MAINPID
+
+[Install]
+WantedBy=multi-user.target
```
redmine/redmine-unicorn.service
上記のと重複してませんか？

RVIRUS0817 commented on redmine/roles/unicorn/files/unicorn.rb:
```diff
@@ -0,0 +1,40 @@
+
+listen "#{app_path}/tmp/unicorn_redmine.sock"
+# プロセスの停止などに必要なPIDファイルの保存先を指定。
+pid "#{app_path}/tmp/unicorn.pid"
```
変数にしなくても相対パスでいけます

RVIRUS0817 commented on redmine/roles/unicorn/files/unicorn.rb:
```diff
@@ -0,0 +1,40 @@
+app_path = "/opt/redmine"
+# Nginxで使用する場合は以下の設定を行う。
+
+listen "#{app_path}/tmp/unicorn_redmine.sock"
```
変数にしなくても相対パスでいけます

RVIRUS0817 commented on redmine/roles/unicorn/files/unicorn.rb:
```diff
@@ -0,0 +1,40 @@
+listen "#{app_path}/tmp/unicorn_redmine.sock"
+# プロセスの停止などに必要なPIDファイルの保存先を指定。
+pid "#{app_path}/tmp/unicorn.pid"
+stderr_path "#{app_path}/log/unicorn.stderr.log"
```
変数にしなくても相対パスでいけます

RVIRUS0817 commented on redmine/roles/unicorn/files/unicorn.rb:
```diff
@@ -0,0 +1,40 @@
+# プロセスの停止などに必要なPIDファイルの保存先を指定。
+pid "#{app_path}/tmp/unicorn.pid"
+stderr_path "#{app_path}/log/unicorn.stderr.log"
+stdout_path "#{app_path}/log/unicorn.stdout.log"
```
変数にしなくても相対パスでいけます

RVIRUS0817 commented on redmine/roles/unicorn/files/unicorn.rb:
```diff
@@ -0,0 +1,40 @@
+stderr_path "#{app_path}/log/unicorn.stderr.log"
+stdout_path "#{app_path}/log/unicorn.stdout.log"
+# Unicornのエラーログと通常ログの位置を指定。
+stderr_path "#{app_path}/log/unicorn.stderr.log"
```
変数にしなくても相対パスでいけます

RVIRUS0817 commented on redmine/roles/unicorn/files/unicorn.rb:
```diff
@@ -0,0 +1,40 @@
+stdout_path "#{app_path}/log/unicorn.stdout.log"
+# Unicornのエラーログと通常ログの位置を指定。
+stderr_path "#{app_path}/log/unicorn.stderr.log"
+stdout_path "#{app_path}/log/unicorn.stdout.log"
```
変数にしなくても相対パスでいけます

RVIRUS0817 commented on redmine/unicorn.rb:
```diff
@@ -0,0 +1,40 @@
+
+after_fork do |server, worker|
+  defined?(ActiveRecord::Base) and ActiveRecord::Base.establish_connection
+end
```
redmine/roles/unicorn/files/unicorn.rb
上記と重複してませんか？

RVIRUS0817 reviewed (commented):
@kazuki-sumi 
コメントしました。

kazuki-sumi committed:
redmine-unicorn重複削除

kazuki-sumi committed:
nginx設定ファイル重複削除

kazuki-sumi committed:
unicorn設定ファイル重複削除

kazuki-sumi committed:
path変更

kazuki-sumi committed:
redmineのダウンロード先変更

kazuki-sumi committed:
railsのバージョン管理を変数化

kazuki-sumi committed:
rubyがインストール済みか確認タスク追加

kazuki-sumi committed:
redmine-menta.meでアクセスできるように名前解決

kazuki-sumi committed:
mysqlのユーザ作成タスク削除

kazuki-sumi committed:
unicornのパスの設定を変更

kazuki-sumi committed:
sockの名称変更

kazuki-sumi committed:
nginxインストールタスクのコード短縮

kazuki-sumi committed:
相対パスに修正

kazuki-sumi commented:
```
PLAY [all] ******************************************************************************************************************

TASK [Gathering Facts] ******************************************************************************************************
ok: [192.168.33.10]

TASK [common : include task/base-package.yml] *******************************************************************************
included: /Users/sumikazuki/Desktop/menta/redmine/roles/common/tasks/base-package.yml for 192.168.33.10

TASK [common : yum install base-package] ************************************************************************************
ok: [192.168.33.10] => (item=vim)
ok: [192.168.33.10] => (item=wget)

TASK [common : include task/host.yml] ***************************************************************************************
included: /Users/sumikazuki/Desktop/menta/redmine/roles/common/tasks/host.yml for 192.168.33.10

TASK [common : name resolution] *********************************************************************************************
ok: [192.168.33.10]

PLAY [redmine] **************************************************************************************************************

TASK [Gathering Facts] ******************************************************************************************************
ok: [192.168.33.10]

TASK [nginx : copy nginx.repo] **********************************************************************************************
ok: [192.168.33.10]

TASK [nginx : install nginx and openssl] ************************************************************************************
ok: [192.168.33.10]

TASK [nginx : start nginx] **************************************************************************************************
ok: [192.168.33.10]

TASK [nginx : copy nginx redmine.conf] **************************************************************************************
changed: [192.168.33.10]

TASK [mysql : delete mariadb] ***********************************************************************************************
ok: [192.168.33.10]

TASK [mysql : install rpm] **************************************************************************************************
ok: [192.168.33.10] => (item=http://dev.mysql.com/get/mysql57-community-release-el7-7.noarch.rpm)

TASK [mysql : install mysqld and mysql-devel] *******************************************************************************
ok: [192.168.33.10] => (item=mysql-community-server)
ok: [192.168.33.10] => (item=mysql-devel)

TASK [mysql : start mysql] **************************************************************************************************
ok: [192.168.33.10]

TASK [ruby : install development tools] *************************************************************************************
ok: [192.168.33.10] => (item=git)
ok: [192.168.33.10] => (item=bzip2)
ok: [192.168.33.10] => (item=gcc)
ok: [192.168.33.10] => (item=openssl-devel)
ok: [192.168.33.10] => (item=readline-devel)
ok: [192.168.33.10] => (item=zlib-devel)
ok: [192.168.33.10] => (item=libyaml-devel)
ok: [192.168.33.10] => (item=libffi-devel)
ok: [192.168.33.10] => (item=gdbm-devel)
ok: [192.168.33.10] => (item=ncurses-devel)
ok: [192.168.33.10] => (item=ImageMagick)
ok: [192.168.33.10] => (item=ImageMagick-devel)

TASK [ruby : git clone rbenv] ***********************************************************************************************
ok: [192.168.33.10]

TASK [ruby : change .rbenv owner] *******************************************************************************************
ok: [192.168.33.10]

TASK [ruby : insert/update /etc/profile env] ********************************************************************************
ok: [192.168.33.10]

TASK [ruby : source /etc/profile] *******************************************************************************************
changed: [192.168.33.10]

TASK [ruby : confirm ruby] **************************************************************************************************
ok: [192.168.33.10]

TASK [ruby : git clone ruby-build] ******************************************************************************************
skipping: [192.168.33.10]

TASK [ruby : install ruby-build] ********************************************************************************************
skipping: [192.168.33.10]

TASK [ruby : install ruby2.5.0] *********************************************************************************************
skipping: [192.168.33.10]

TASK [ruby : set ruby2.5.0 to gloval] ***************************************************************************************
skipping: [192.168.33.10]

TASK [rails : install bundler] **********************************************************************************************
changed: [192.168.33.10]

TASK [rails : install rails 5.2.2] ******************************************************************************************
changed: [192.168.33.10]

TASK [redmine : change permision of var] ************************************************************************************
ok: [192.168.33.10]

TASK [redmine : mkdir www] **************************************************************************************************
ok: [192.168.33.10]

TASK [redmine : confirm redmine] ********************************************************************************************
ok: [192.168.33.10]

TASK [redmine : download redmine] *******************************************************************************************
skipping: [192.168.33.10]

TASK [redmine : unpack redmine] *********************************************************************************************
skipping: [192.168.33.10]

TASK [redmine : rename from redmine-4.0.0 to redmine] ***********************************************************************
skipping: [192.168.33.10]

TASK [redmine : copy Gemfile] ***********************************************************************************************
ok: [192.168.33.10]

TASK [redmine : copy database.yml] ******************************************************************************************
ok: [192.168.33.10]

TASK [redmine : bundle lock x86-mingw32 x64-mingw32 x86-mswin32] ************************************************************
changed: [192.168.33.10]

TASK [redmine : create /var/www/redmine/app/assets/config] ******************************************************************
ok: [192.168.33.10]

TASK [redmine : create manifest.js] *****************************************************************************************
ok: [192.168.33.10]

TASK [unicorn : copy unicorn.rb] ********************************************************************************************
changed: [192.168.33.10]

TASK [unicorn : copy redmine-unicorn.service] *******************************************************************************
ok: [192.168.33.10]

TASK [unicorn : disable SELinux] ********************************************************************************************
ok: [192.168.33.10]

TASK [unicorn : start unicorn] **********************************************************************************************
ok: [192.168.33.10]

RUNNING HANDLER [nginx : restart nginx] *************************************************************************************
changed: [192.168.33.10]

PLAY RECAP ******************************************************************************************************************
192.168.33.10              : ok=35   changed=7    unreachable=0    failed=0    skipped=7    rescued=0    ignored=0   

```

kazuki-sumi commented:
<img width="1280" alt="スクリーンショット 2020-02-20 18 31 23" src="https://user-images.githubusercontent.com/38270446/74920339-5d03e380-540f-11ea-84d0-502fa3ac03f7.png">

RVIRUS0817 commented:
@kazuki-sumi 
良いと思います！
